metabox/README.md

```diff
@@ -0,0 +1,84 @@
+
+Metabox comes with **scenarios** to test different parts of Checkbox. A scenario is a Python package that defines what to run, and what is expected to happen. For instance, the [`desktop_env` scenario] prepares a launcher with a simple test plan containing desktop jobs, executes it and makes sure the jobs pass and their output contain some expected information.
+
+Users define **configurations** to execute these scenarios. A configuration is a Python module that contains what version(s) of Checkbox must be tested, and on what Ubuntu release(s). Some configurations can be found in the [`configs` directory], but users can write their own.
+
+# Installation
+
+Metabox requires LXD. Check the [LXD documentation to install and initialize it].
+
+Create a Python virtual environment and install Metabox in it:
+
+```shell
+$ cd metabox/
+$ python3 -m venv metabox
+$ source metabox/bin/activate
+(metabox) $ pip install -e .
+```
+
+# Usage
+
+```
+usage: metabox [-h] [--tag TAGS] [--exclude-tag EXCLUDE_TAGS] [--log {TRACE,DEBUG,INFO,SUCCESS,WARNING,ERROR,CRITICAL}] [--do-not-dispose] [--hold-on-fail] [--debug-machine-setup] CONFIG
+
+positional arguments:
+  CONFIG                Metabox configuration file
+
+optional arguments:
+  -h, --help            show this help message and exit
+  --tag TAGS            Run only scenario with the specified tag. Can be used multiple times.
+  --exclude-tag EXCLUDE_TAGS
+                        Do not run scenario with the specified tag. Can be used multiple times.
+  --log {TRACE,DEBUG,INFO,SUCCESS,WARNING,ERROR,CRITICAL}
+                        Set the logging level
+  --do-not-dispose      Do not delete LXD containers after the run
+  --hold-on-fail        Pause testing when a scenario fails
+  --debug-machine-setup
+                        Turn on verbosity during machine setup. Only works with --log TRACE
+```
+
+Let's say I want to test:
+
+- the [`basic` scenario] (which focuses on Checkbox local)
+- using the latest Debian version of Checkbox available in the Daily Builds PPA
+- on bionic (18.04) and focal (20.04)
+- for Checkbox local only (not Checkbox remote)
+
+I can create the following `local-daily-builds-config.py` file:
+
+```python
+configuration = {
+    'local': {
+        'origin': 'ppa',
+        'uri': 'ppa:checkbox-dev/ppa',
+        'releases': ['bionic', 'focal'],
+    },
+}
+```
+
+Then call Metabox with it:
+
+```
+$ metabox local-daily-builds-config.py --do-not-dispose --tag basic
+```
+
+- `--tag basic` will ensure only scenarios that contains the term `basic` will be run.
+- `--do-not-dispose` prevents Metabox from deleting the Linux containers it created. This will save you tons of time, since Metabox will only download the required image and setup the container once, create a snapshot, then stop the container once the testing is done. The next time you run the command, Metabox will reopen the existing container and rollback to a clean state in it before starting the new tests.
```

Comment:
The goal of do-not-dispose is to leave a container that hosted a failing test as is, so operator can log back in to it and investigate the crime scene.

Reply:
I think you mean either `--hold-on-fail` or `--debug-machine-setup`. If I remove `--do-not-dispose`, the LXD containers are deleted after the tests are run:

```
$ metabox local-config.py --log TRACE --tag basic
13:42:43 | DEBUG    Registering a scenario: basic.run-invocation.RunTestplan
13:42:43 | DEBUG    Registering a scenario: basic.run-invocation.RunFailingTestplan
13:42:43 | DEBUG    Registering a scenario: basic.run-invocation.RunTestplanWithEnvvar
13:42:43 | DEBUG    Registering a scenario: basic.run-invocation.RunManualplan
13:42:43 | DEBUG    Registering a scenario: desktop_env.launcher.GlxGears
13:42:43 | DEBUG    Registering a scenario: desktop_env.launcher.AudioPlayback
13:42:43 | DEBUG    Registering a scenario: restart.launcher.Reboot
13:42:43 | DEBUG    Registering a scenario: urwid.testplan.UrwidTestPlanSelection
13:42:43 | DEBUG    Skipping a scenario: [remote] desktop_env.launcher.GlxGears
13:42:43 | DEBUG    Skipping a scenario: [remote] desktop_env.launcher.AudioPlayback
13:42:43 | DEBUG    Skipping a scenario: [remote] restart.launcher.Reboot
13:42:43 | INFO     Including scenario tag(s): basic
13:42:43 | DEBUG    Combo: {<local alias:'focal' origin:'ppa'>}
13:42:48 | DEBUG    snap LXD profile updated successfully
13:42:48 | DEBUG    lowmem LXD profile updated successfully
13:42:48 | DEBUG    checkbox LXD profile updated successfully
13:42:48 | DEBUG    Getting information about metabox-local-focal-ppa...
13:42:50 | DEBUG    Config: <local alias:'focal' origin:'ppa'>
13:42:50 | DEBUG    Adding <local alias:'focal' origin:'ppa'> to list of owned containers...
13:42:50 | DEBUG    [restored    ] metabox-local-focal-ppa
13:42:52 | DEBUG    [initializing] metabox-local-focal-ppa
13:42:53 | DEBUG    [starting    ] metabox-local-focal-ppa
13:42:58 | DEBUG    [running     ] metabox-local-focal-ppa
13:43:00 | INFO     Starting scenario: basic.run-invocation.RunFailingTestplan
(...)
13:43:33 | SUCCESS  [local][('focal',)] basic.run-invocation.RunManualplan scenario has passed.
13:43:38 | DEBUG    [stopped     ] metabox-local-focal-ppa
13:43:38 | DEBUG    [deleted     ] metabox-local-focal-ppa
--------------------------------------------------------------------------------
13:43:38 | SUCCESS  Ran 3 scenarios in 48.814s


$ lxc list
To start your first container, try: lxc launch ubuntu:22.04
Or for a virtual machine: lxc launch ubuntu:22.04 --vm

+------+-------+------+------+------+-----------+
| NAME | STATE | IPV4 | IPV6 | TYPE | SNAPSHOTS |
+------+-------+------+------+------+-----------+
```

Reply:
So many options :D Thanks for the info!